lib/Sema/InstrumenterSupport.h

```diff
@@ -0,0 +1,152 @@
+using namespace swift;
+
+namespace swift {
+namespace instrumenter_support {
+
+template <class E> class Added {
+private:
+  E Contents;
+public:
+  Added() { }
+  Added(E NewContents) { Contents = NewContents; }
+  const Added<E> &operator=(const Added<E> &rhs) {
+    Contents = rhs.Contents;
+    return *this;
+  }
+  E &operator*() { return Contents; }
+  E &operator->() { return Contents; }
+};
+
+class InstrumenterBase {
+
+protected:
+  InstrumenterBase() : CF(*this) {}
+  virtual ~InstrumenterBase() = default;
+  virtual void anchor();
+  virtual BraceStmt *transformBraceStmt(BraceStmt *BS,
+                                        bool TopLevel = false) = 0;
+
+  class ErrorGatherer : public DiagnosticConsumer {
+  private:
+    bool error = false;
+    DiagnosticEngine &diags;
+  public:
+    ErrorGatherer(DiagnosticEngine &diags) : diags(diags) {
+      diags.addConsumer(*this);
+    }
+    ~ErrorGatherer() {
+      diags.takeConsumers();
+    }
+    virtual void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                                  DiagnosticKind Kind, StringRef Text,
+                                  const DiagnosticInfo &Info) {
+      if (Kind == swift::DiagnosticKind::Error) {
+        error = true;
+      }
+      llvm::errs() << Text << "\n";
+    }
+    bool hadError() { 
+      return error;
+    }
+  };
+
+  class ClosureFinder : public ASTWalker {
+  private:
+    InstrumenterBase &I;
+  public:
+    ClosureFinder (InstrumenterBase &Inst) : I(Inst) { }
+    virtual std::pair<bool, Stmt*> walkToStmtPre(Stmt *S) {
+      if (isa<BraceStmt>(S)) {
+        return { false, S }; // don't walk into brace statements; we
+                             // need to respect nesting!
+      } else {
+        return { true, S };
+      }
+    }
+    virtual std::pair<bool, Expr*> walkToExprPre(Expr *E) {
+      if (ClosureExpr *CE = dyn_cast<ClosureExpr>(E)) {
+        BraceStmt *B = CE->getBody();
+        if (B) {
+          BraceStmt *NB = I.transformBraceStmt(B);
+          CE->setBody(NB, false);
+          // just with the entry and exit logging this is going to
+          // be more than a single expression!
+        }
+      }
+      return { true, E };
+    }
+  };
+    
+  ClosureFinder CF;
+
+  class ErrorFinder : public ASTWalker {
+    bool error = false;
+  public:
+    ErrorFinder () { }
+    virtual std::pair<bool, Expr*> walkToExprPre(Expr *E) {
+      if (isa<ErrorExpr>(E) || !E->getType() || E->getType()->hasError()) {
+        error = true;
+        return { false, E };
+      }
+      return { true, E };
+    }
+    virtual bool walkToDeclPre(Decl *D) {
+      if (ValueDecl *VD = dyn_cast<ValueDecl>(D)) {
+        if (!VD->hasInterfaceType() ||
+            VD->getInterfaceType()->hasError()) {
+          error = true;
+          return false;
+        }
+      }
+      return true;
+    }
+    bool hadError() { 
+      return error;
+    }
+  };
+
+  template <class T> bool doTypeCheck(ASTContext &Ctx,
```

Comment:
I can't seem to figure out how to do this. If the Impl's type is:
```
bool InstrumenterBase::doTypeCheckImpl(ASTContext &Ctx, DeclContext *DC,
                                       Added<Expr *> &parsedExpr)
```
Then the code would be:
```
  Expr *E = *parsedExpr;
  TC.typeCheckExpression(E, DC);
  parsedExpr = Added<Expr *>(E);

  if (*parsedExpr) {
    ErrorFinder errorFinder;
    parsedExpr->walk(errorFinder);
```
Wouldn't this fail because, without the T, the `walk` call here call `Expr *`'s `walk` instead of `T`'s `walk`?